src/cls/2pc_queue/cls_2pc_queue.cc

```diff
@@ -594,6 +605,17 @@ static int cls_2pc_queue_remove_entries(cls_method_context_t hctx, bufferlist *i
   if (ret < 0) {
     return ret;
   }
+
+  // Old RGW is running, and it sent cls_queue_remove_op instead of cls_2pc_queue_remove_op
+  if (rem_2pc_op.entries_to_remove == 0) {
+    CLS_LOG(10, "INFO: cls_2pc_queue_remove_entries: incompatible RGW with rados, counting entries to remove...");
+    cls_queue_list_op list_op;
+    list_op.max = std::numeric_limits<uint64_t>::max(); // max length because endmarker is the stopping condition.
+    list_op.end_marker = rem_2pc_op.end_marker;
+    rem_2pc_op.entries_to_remove = cls_2pc_queue_count_entries(hctx, list_op, head);
```

Comment:
It might be true that in most cases we would like to remove all the entries after push endpoint succeeds, however, if we have an expiry config (say 10 seconds), at time t+10 we will want to remove only the entries that were added at time t.

Reply:
am i correct that `entries_to_remove` is only used to update the queue stats? why would we need the _client_ to send that, instead of tracking it on the server side? the client can't know how many entries the server will remove if other clients are racing to remove things too

if the server (cls) handled this internally, we wouldn't need to change anything on the client side so backward compat is far simpler

Reply:
I don't think that there are multiple clients racing to remove entries, only one RGW owns a queue at any point of time.

We remove the entries by making  `start_marker` will point to  `end_marker` we send:
```
head.front = end_marker;
```
so we don't iterate on the entries we remove, although, maybe we can calculate the number of entries we remove by:
```
uint64_t len = head.queue_size - head.front.offset;
```
what do you think @yuvalif?

Reply:
having the client send the number of entries to remove is a bit hacky.
but, as Ali wrote, would work correctly with the existing implementation and is the most efficient way to implement that.

Reply:
our other queues handle racing writers correctly, so it would be nice to preserve that here. be careful about baking too many assumptions from rgw's notifications into cls_2pc_queue, or it won't be suitable for other uses

though if we're trying to replace this with fifo, maybe we don't care much about the future of cls_2pc_queue?

Reply:
@cbodle how about the following option:
in cls_2pc_queue_ops.h add a comment saying:
"if no race condition is possible while deleting entries from the queue (as with our current implementation where only one RGW locks the queue and delete entries from it), use ver2 of `cls_2pc_queue_remove_op`. if multiple clients may be deleting entries from the queue, use ver1 of `cls_2pc_queue_remove_op`."

Reply:
@AliMasarweh lets go with this option. just make sure that v1 and v2 are documented in the client header file

Reply:
I think we have another approach, if we save in the header of the queue (`cls_2pc_urgent_data`) the entry size, we can calculate how entries are removed, because we can tell the size that we are removing from 
```
int queue_remove_entries(cls_method_context_t hctx, const cls_queue_remove_op& op, cls_queue_head& head)
```
I am still not 100% sure that this is possible, since most of the times the entry size is 519, in some times it will be 518 or 520. so I might need to do some padding to the entry or do an estimation to get the correct number.
Do you think I should investigate this idea? @yuvalif @cbodley

Reply:
@AliMasarweh i see 2 issues with this idea:
* this does not fix the upgrade issue, since the entry sizes won't exist in the header
* each entry may have a different size. you would need to save a map of all sizes with their offsets, and do the calculation based on that